### PR TITLE
Add thai2rom algorithm

### PR DESCRIPTION
**Pull request by waf**
Adds the Thai2Rom algorithm, originally from [PyThaiNLP](https://github.com/PyThaiNLP/pythainlp/blob/dev/pythainlp/transliterate/thai2rom.py).

As part of this, I needed to move the existing TLTK algorithm to a subdirectory. That's the first commit and doesn't need much review. The second commit is the actual implementation of the Thai2Rom algorithm, using TorchSharp.

Interesting parts of this PR:

- Each algorithm now has its own README, as the requirements and development guides are very different. There's still a README at the root of the repository with general usage and introduction material.
- I added a `ThaiRomanizationSharp.Abstractions` project, which contains a shared interface that both projects implement.
- While the external interface for users follows C# standard naming conventions, the internal code of Thai2Rom does not, as it is closely mimicking the Python code. I tended to follow Python conventions more than C# conventions, because it will be easier to upgrade in the future. See the Development Guide portion of the Thai2Rom README for more information.

### CODE REVIEW

File: src/ThaiRomanizationSharp.Thai2Rom/Models/Loader.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+        public static ModelParameters LoadParameters()
+        {
+            var parameters = JsonSerializer.Deserialize<ModelParameters>(
+                File.ReadAllText(ModelParameterPath)
+            );
+
+            return parameters is not null
+                ? parameters
+                : throw new InvalidOperationException($"Unable to load required {ModelParameterPath}");
```
Comment (aaronamm) [inline src/ThaiRomanizationSharp.Thai2Rom/Models/Loader.cs:23]: I miss a format like this.

File: src/ThaiRomanizationSharp.Thai2Rom/NeuralNet/Attn.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            :base("attn")
+        {
+            this.method = method;
+            this.hiddenSize = hiddenSize;
+
+            if(this.method == "general")
+            {
+                this.attn = nn.Linear(this.hiddenSize, hiddenSize);
+            }
+            else// if(this.method == "concat")
```
Comment (aaronamm) [inline src/ThaiRomanizationSharp.Thai2Rom/NeuralNet/Attn.cs:24]: Very minor, please remove this code comment.

File: src/ThaiRomanizationSharp.Thai2Rom/NeuralNet/Attn.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                ); // (batch_size * sequence_len,  hidden_size)
+                attnEnergies = attnEnergies
+                    .view(encoderOutputs.size())
+                    .bmm(
+                        hidden.transpose(1, 2))
+                    .squeeze(
+                        2
+                    );
+            }
+            else// if (this.method == "concat")
```
Comment (aaronamm) [inline src/ThaiRomanizationSharp.Thai2Rom/NeuralNet/Attn.cs:53]: Not sure you have intension to keep this statement `if (this.method == "concat")
Write`?

File: src/ThaiRomanizationSharp.Thai2Rom/NeuralNet/Seq2Seq.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+using System.Diagnostics;
+using System.Linq;
+using TorchSharp;
+using static TorchSharp.torch;
+
+namespace ThaiRomanizationSharp.Thai2Rom.NeuralNet
+{
+    using static ThaiRomanizationSharp.Thai2Rom.Extensions;
+
+    internal sealed class Seq2Seq : nn.CustomModule
```
Comment (aaronamm) [inline src/ThaiRomanizationSharp.Thai2Rom/NeuralNet/Seq2Seq.cs:11]: Could you please give just a short summary of what this class does/for?
Does Seq2Seq stand for `Sequence to Sequence`?

File: src/ThaiRomanizationSharp.Thai2Rom/Thai2RomService.cs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+using ThaiRomanizationSharp.Thai2Rom.NeuralNet;
+using TorchSharp;
+using static TorchSharp.torch;
+
+namespace ThaiRomanizationSharp.Thai2Rom
+{
+    using static ThaiRomanizationSharp.Thai2Rom.Extensions;
+
+    /// <inheritdoc cref="IThaiRomanizationService"/>
+    public sealed class Thai2RomService : IThaiRomanizationService
```
Comment (aaronamm) [inline src/ThaiRomanizationSharp.Thai2Rom/Thai2RomService.cs:16]: Thanks for making an abstract class IThaiRomanizationService

File: src/ThaiRomanizationSharp.Thai2Rom/Thai2RomService.cs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+            this.network.load("Models/Data/model_state_dict");
+            this.network.Eval();
+        }
+
+        /// <inheritdoc cref="IThaiRomanizationService.Romanize(string)"/>
+        public string Romanize(string text)
+        {
+            if (text is null) throw new ArgumentNullException(nameof(text));
+            if (text is "") return "";
```
Comment (aaronamm) [inline src/ThaiRomanizationSharp.Thai2Rom/Thai2RomService.cs:57]: At first I have thought if `text is string.empty` would work.

File: test/ThaiRomanizationSharp.Thai2Rom.Test/Thai2RomServiceTest.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        public ThaiRomanizationServiceTest() =>
+            this.thai2RomService = new Thai2RomService();
+
+        [Theory]
+        // simple thai
+        [InlineData("สวัสดี", "sawatdi")]
+        [InlineData("นะจ๊ะ", "nacha")]
+        [InlineData("ไม่รู้ ไม่รู้", "mairu mairu")]
+        // thai intermixed with other languages. Should leave the other languages alone.
+        [InlineData("สวัสดี Hello こんにちは สวัสดีตอนเย็น", "sawatdi Hello こんにちは sawatditonyen")]
```
Comment (aaronamm) [inline test/ThaiRomanizationSharp.Thai2Rom.Test/Thai2RomServiceTest.cs:19]: Good test case, thanks for adding it.

File: test/ThaiRomanizationSharp.Tltk.Test/ThaiRomanizationServiceTest.cs [hunk 8]
```diff
@@ -19,11 +20,14 @@ public ThaiRomanizationServiceTest(ITestOutputHelper output)
...
-            var actualOutputText = thaiRomanizationService.ToRoman(inputText);
+            var actualOutputText = thaiRomanizationService.Romanize(inputText);
             // There is a output if you run a test with "Run Test" code lens but not .NET Core Test Explorer
             output.WriteLine($"expectedOutputText: {actualOutputText}");
 
             Assert.Equal(expectedOutputText, actualOutputText);
         }
+
+        public void Dispose() =>
+            thaiRomanizationService.Dispose();
```
Comment (aaronamm) [inline test/ThaiRomanizationSharp.Tltk.Test/ThaiRomanizationServiceTest.cs:31]: Thanks for adding this, I forgot to dispose it.

**Review by aaronamm — CHANGES_REQUESTED** [review]
Very good PR.
- Thank for a details read me.
- I have only minor feedbacks but wonder for some code comment if you intend to  keep them.
- I don't check in details of implementation because I don't have an idea about it just normal coding review.
- I would like to change some code to Lambda function but I think we want to keep this because it will be easy for use to update code when there is new update form original Python code.

Thank you so much for your hard working.

Next we can:
- Create GitHub Actions workflow to run our unit test.
- Publish to NugetPackage
- Create a simple website to demo this tranlation.

Thanks.

File: src/ThaiRomanizationSharp.Thai2Rom/NeuralNet/Attn.cs (hunk 3 above)
Comment (waf) [reply src/ThaiRomanizationSharp.Thai2Rom/NeuralNet/Attn.cs:53]: I do intend to keep it; it's part of the python code that I didn't need. I'll update this to be more obvious, thanks!

File: src/ThaiRomanizationSharp.Thai2Rom/Thai2RomService.cs (hunk 6 above)
Comment (waf) [reply src/ThaiRomanizationSharp.Thai2Rom/Thai2RomService.cs:57]: Yep, only literals in patterns 🚮

**Review by aaronamm — APPROVED** [review]

**Comment by aaronamm** [comment]
Thank you for your contribution. I appreciate it.

**Pull request merged**